### ISSUE TITLE
Controller: Simplify timeout management for transactions

### ISSUE THREAD

**Issue by pbelgundi** [pravega/pravega#6241]
**Is your feature request related to a problem? Please describe.**
When a Pravega Transaction is created, users set a "lease" time for the Transaction. If the transaction does not close _(commit or abort)_ within the provided lease interval from creation time it is eagerly aborted by Controller. To avoid this, a user application _(Pravega Client)_ makes pingTxn API() calls to Controller to extend the initial "lease" period periodically. More than 1 pingTxn calls are made per lease interval based on a certain formula with exponential retries ( 1 for <4s, 2 for <9s, 3 for <16s, 4 for <25s, ... 10 for <100s).

This design has the following problems:
a. All open transactions are aborted when a user application goes down for an extended period ( > few mins) as pings are not sent for any open transaction.
b. Generates unnecessary load of additional traffic from Client to Controller (_pingTxn() calls need to be made for every transaction and multiple calls are made per "lease" interval_) The `lease` time for a transaction can only be extended by a few minutes by every `pingTxn` API call thus resulting in a very large number of pingTxn calls being made to Controller when transaction processing slows down.
c. Increased commit and create transaction processing delay, as the transaction needs to be added to and removed from the timeout service that keeps track of transaction timeouts.
d. If pingTxn API calls fail because of a network issue txns get aborted unnecessarily and have to be restarted.
e. The current Transaction Timeout Service holds all transaction data in memory causing a lot of load on Controller memory, esp when number of transactions in open state is high, often causing the Controller to go OOM.

**Describe the solution you'd like**
- It is important to "rethink" the need for Transaction timeouts in Pravega and their design. We do not require to micro-manage a transactions' lifetime, as we **do not** stall reads or writes on the original Stream for our transaction processing.

-  Unlike a lot of other streaming systems, events written to a Transactions are not added directly to the source Stream. Instead they written to a separate set of segments that get merged into the source Stream on commit. The intent for having _timeouts_ in the other streaming systems is to unblock "readers" who can stall because an open transaction. However, Pravega does not have this limitation and readers can still continue reading from the source Stream while transactional writes are in progress. 

- All we need is a way to **reclaim** resources if a transaction was opened but not written to _(abandoned)_ by the user. 

**Suggested Solution**
- Remove the current transaction timeout service that actively tracks each transaction and aborts it as soon as the txn has lived past its lease period. 

-  Create a new _Transaction GC Service_ that would run & abort all transactions that are beyond their configured lifetime. This service would run periodically (default interval could be 60 mins) and the run interval would be a configurable value to be set on Controller startup.  In case of application downtime, this service could be **disabled**, so open transactions beyond their lease period would **not** be aborted till the service is restarted and runs next. 

- The upper bound for transaction lease time would be LONG_MAX. A user can set a transaction lease value to LONG_MAX if he never wants it to be garbage collected. Note that these transactions would not need any pings to be kept alive and will never be aborted by the GC service.

- The lower bound for transaction lease would be 2 minutes _(currently this is 10 secs)_. 

- pingTxn API to be removed from Controller. Pinger Service to be removed from Client. Client should not make any pingTxn calls to Controller. The Client does not have any information about weather a transactions' lease should be extended or not and today it blindly makes pingTxn() API calls to keep extending the lease. Instead it may be better for the Client to do this when requested by the user application as the application would have knowledge about weather the transaction needs to be kept open or not.

- A new API extendTransactionLease(long extendms) API to be added to Controller and Client. A user application wanting to extend the lease of a transaction should explicitly invoke this API on Client (Client internally invokes this API on Controller) and it would result in the lifetime of a transaction being extended.
For extendms min value = 2 minutes and max value = 525600

**Comment by tkaitchuck** [comment]
> a. All open transactions are aborted when a user application goes down for an extended period ( > few mins) as pings are not sent for any open transaction.

This isn't so much a "problem" but rather the entire goal of the feature.

> Create a new Transaction GC Service that would run & abort all transactions that are beyond their configured lifetime. This service would run periodically (default interval could be 60 mins) and the run interval would be a configurable value to be set on Controller startup. In case of application downtime, this service could be disabled, so open transactions beyond their lease period would not be aborted till the service is restarted and runs next.

Wouldn't that mean that transactions effectively can't have a working timeout of less than 60 minutes?

**Comment by tkaitchuck** [comment]
When would the transaction segments actually be sealed?

**Comment by pbelgundi** [comment]
> This isn't so much a "problem" but rather the entire goal of the feature.

This is a huge problem for flink today and spark actually works around it by setting larger lease values.

**Comment by pbelgundi** [comment]
> > a. All open transactions are aborted when a user application goes down for an extended period ( > few mins) as pings are not sent for any open transaction.
> 
> This isn't so much a "problem" but rather the entire goal of the feature.
> 
> > Create a new Transaction GC Service that would run & abort all transactions that are beyond their configured lifetime. This service would run periodically (default interval could be 60 mins) and the run interval would be a configurable value to be set on Controller startup. In case of application downtime, this service could be disabled, so open transactions beyond their lease period would not be aborted till the service is restarted and runs next.
> 
> Wouldn't that mean that transactions effectively can't have a working timeout of less than 60 minutes?

Yes it would if your GC Service interval is 60 mins. It should not be concern because it does not stall readers/writers from operating on the original Stream.

**Comment by RaulGracia** [comment]
If I remember correctly, instead of redesigning the timeouts for Transactions, we identified several points that can be improved to the current scheme instead (i.e., the formula for limiting or reducing the number of pings can be fixed to do less number of pings per txn lease period. One suggestion was to divide the lease interval by a certain number and send as many pings per lease interval.). 

We can create a separate issue specifically for that.
